### ISSUE TITLE
Currently facing issue in catalog search using search with SKU keyword

### ISSUE THREAD

**Issue by sw-tt-jigneshvagh** [magento/magento2#31466]
### Preconditions (*)

1. I am usng magento 2.4.0 version
2. I am using magento-cloud started plan
3. Reproducible on clean 2.4-develop
### Steps to reproduce (*)

1. In Products, I have "170000D000237" SKU named product, but when I will search from Frontend Area from topsearch with some SKU keywords like "D000237" its giving "Your search returned no results." message (not return data). but from first letters like "170000" its giving me proper data. So please tell me what i am doing wrong here.

2. In same case other product I have like "C-09700-QSJA2" its also behave as above I mentioned.

### Expected result (*)

1. In above two case I mentioned it will give me not found result, but it should be give me the result data in all cases which keywords I will search with from SKU, Name from it I will type from last charactor or between-to-last charactors in it.

### Actual result (*)

1. Found Data 
![returned-data](https://user-images.githubusercontent.com/25668192/103165197-94ff6280-483a-11eb-922c-eca2362dbf7f.jpg)

2. Not Found Data 
![not-return-data-2](https://user-images.githubusercontent.com/25668192/103165053-eb6ba180-4838-11eb-81e5-1552e57501c5.jpg)
![not-return-data-1](https://user-images.githubusercontent.com/25668192/103165051-ead30b00-4838-11eb-8281-62d72b8b460f.jpg)
---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [x] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by engcom-Bravo** [comment]
Hello @sw-tt-jigneshvagh 

Thank you for your report.

The issue is reproducible on clean 2.4-develop using _quick search_
![quick_search](https://user-images.githubusercontent.com/51680745/103215871-4b3b7880-491d-11eb-8fd3-0bd79bd5a3b8.png)

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40245` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40245` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Bravo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-40245` were created

**Issue Available**: @engcom-Bravo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by sw-tt-jigneshvagh** [comment]
any updates on this @engcom-Bravo @magento

**Comment by johebting** [comment]
any updates on this ?

**Comment by gowrizrh** [comment]
@johebting @sw-tt-jigneshvagh I'm not sure why this is confirmed as an issue by the Magento team but

From the looks of it, this is the way the Elasticsearch module for Magento works.

In the catalog search module there's an explicit Elasticsearch analyzer defined to search on prefixes.

```xml
<!-- vendor/magento/module-catalog-search/etc/di.xml:276 -->

<type name="Magento\CatalogSearch\Model\Search\Request\MatchQueriesModifier">
    <arguments>
        <argument name="queries" xsi:type="array">
            <item name="partial_search" xsi:type="array">
                <item name="name" xsi:type="array">
                    <item name="analyzer" xsi:type="string">prefix_search</item>
                </item>
                <item name="sku" xsi:type="array">
                    <item name="analyzer" xsi:type="string">sku_prefix_search</item>
                </item>
            </item>
        </argument>
    </arguments>
</type>
```

I'm using the sample data to search for the product SKU `WH01`


Searching `WH0` will return search results

![image](https://user-images.githubusercontent.com/40108018/188772893-a3a2c57f-8fc5-4b23-bc48-d34549b2e3e0.png)

Searching `H01` will return nothing

![image](https://user-images.githubusercontent.com/40108018/188773040-43ba59d1-7b4a-41ee-ac57-a90d83efbd02.png)



If you simply take the raw request that Magento performs on the product index in Magento, you can see the prefix clause there.
![image](https://user-images.githubusercontent.com/40108018/188773144-902977b7-0739-4272-ab43-ed15585d7ad5.png)

Note that you can perform wildcard queries on document fields in Elasticsearch but that is not to be confused with the prefix analyzer defined in Magento.

Example: You can construct queries that could use wildcards
![image](https://user-images.githubusercontent.com/40108018/188773607-b5db55c9-d95c-48d0-be8f-7bb9b1972b5b.png)

In conclusion this is not an issue.

**Comment by johebting** [comment]
This may not be a _technical_ problem since the behavior matches the ES query, but you can't just assume that this behavior matches a product reference search.

In my opinion, this search should work regardless of how the catalog and SKUs are constructed.

If for some reason the catalog is built with SKUs following this pattern: `{color}-{size}-{parent_reference}` , how can anyone expect to get relevant search results if Magento or ES only search in the prefix?

Maybe Magento can keep this behavior, even as the default one, but propose others via a new configuration of the search modules.

**Comment by gowrizrh** [comment]
The search isn't only in the prefix. The prefix is an additional match on all the existing rules.

Personally I don't see a case where a customer or a system would search partially by SKUs given every other attribute is also indexed.

The prefix is mostly a convenience feature and it is up to Magento to decide if this is a feature that should be supported in future.

**Comment by engcom-Bravo** [comment]
@magento give me 2.4-develop instance

**Comment by engcom-Bravo** [comment]
Hi @sw-tt-jigneshvagh,

Thanks for your reporting and collaboration.

We have verified the issue and it seems to be an expected behaviour of magento.Hence we are considering this as a improvement and we are considering this as feature request.

Thanks.
